projects/core/src/auth/user-auth/services/oauth-lib-wrapper.service.ts

```diff
@@ -111,11 +113,37 @@ export class OAuthLibWrapperService {
 
   /**
    * Tries to login user based on `code` or `token` present in the url.
+   *
+   * @param result The result returned by `OAuthService.tryLogin()`.
+   *
+   * @param tokenReceived Whether the event 'token_received' is emitted during `OAuthService.tryLogin()`.
+   * We can use this identify that we have returned from an external authorization page to Spartacus).
+   * In cases where we don't receive this event, the token has been obtained from storage.
    */
-  tryLogin() {
-    return this.oAuthService.tryLogin({
-      // We don't load discovery document, because it doesn't contain revoke endpoint information
-      disableOAuth2StateCheck: true,
+  tryLogin(): Promise<OAuthTryLoginResult> {
+    return new Promise((resolve) => {
+      // We use the 'token_received' event to check if we have returned
+      // from the auth server.
+      let tokenReceivedEvent: OAuthEvent | undefined;
+      const subscription = this.events$
+        .pipe(
+          filter((event) => event.type === 'token_received'),
+          take(1)
+        )
+        .subscribe((event) => (tokenReceivedEvent = event));
+
+      this.oAuthService
+        .tryLogin({
+          // We don't load discovery document, because it doesn't contain revoke endpoint information
+          disableOAuth2StateCheck: true,
+        })
+        .then((result: boolean) => {
+          subscription.unsubscribe();
```

Comment:
We should unsubscribe regardless `oAuthService.tryLogin` resolves or **rejects** the promise (e.g. due to a runtime error).
So let's move `subscription.unsubscribe();` to the chained method `.finally()` of the promise.